source/common/router/router.cc

```diff
@@ -211,18 +211,24 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Determine if there is a direct response for the request.
-  if (route_->directResponseEntry()) {
-    auto response_code = route_->directResponseEntry()->responseCode();
-    if (response_code >= Http::Code::MultipleChoices && response_code < Http::Code::BadRequest) {
-      config_.stats_.rq_redirect_.inc();
-      Http::Utility::sendRedirect(*callbacks_, route_->directResponseEntry()->newPath(headers),
-                                  response_code);
-      return Http::FilterHeadersStatus::StopIteration;
-    }
+  const auto* direct_response = route_->directResponseEntry();
+  const auto& request_headers = headers;
+  if (direct_response != nullptr) {
     config_.stats_.rq_direct_response_.inc();
-    sendLocalReply(route_->directResponseEntry()->responseCode(),
-                   route_->directResponseEntry()->responseBody(), false);
-    // TODO(brian-pane) support sending response_headers_to_add.
+    Http::Utility::sendLocalReply(
+        [this, direct_response, &request_headers](Http::HeaderMapPtr&& headers,
+                                                  bool end_stream) -> void {
+          const auto new_path = direct_response->newPath(request_headers);
+          if (!new_path.empty()) {
+            headers->addCopy(Http::LowerCaseString("location"), new_path);
```

Comment:
You can do this using O(1) headers by doing a variant of `headers->insertLocation()...`

Reply:
It looks like Location isn't configured as one of the O(1) headers in `header_map.h`. I suppose I could add it, but it looks like each new header configured for O(1) access adds a pointer to `HeaderMapImpl` ... is it okay to increase the size of that structure?

Reply:
Ah OK, sorry. It's not a big deal either way. But you can use the constant that is already in `Http::Headers` and then `addReferenceKey`